Manifold/Type.swift

```diff
@@ -46,30 +211,54 @@ public enum Type: Hashable {
 	public var hashValue: Int {
 		return analysis(
 			ifVariable: { $0.hashValue },
-			ifFunction: { $0.hashValue ^ $1.hashValue },
+			ifConstructed: { $0.hashValue },
 			ifUniversal: { $0.hashValue ^ $1.hashValue }
 		)
 	}
-}
-
 
-/// Equality defined up to renaming.
-public func == (left: Type, right: Type) -> Bool {
-	switch (left, right) {
-	case (.Variable, .Variable):
-		return true
 
-	case let (.Function(x1, x2), .Function(y1, y2)):
-		return x1.value == y1.value && x2.value == y2.value
+	// MARK: Printable
 
-	case let (.Universal(a1, t1), .Universal(a2, t2)):
-		return a1 == a2 && t1 == t2
+	public var description: String {
+		return describe()
+	}
 
-	default:
-		return false
+	private func describe(_ boundVariables: Set<Manifold.Variable> = []) -> String {
+		let bound = "α"
+		let free = "τ"
+		return analysis(
+			ifVariable: { (boundVariables.contains($0) ? bound : free) + $0.value.subscriptedDescription },
+			ifConstructed: { c in
+				c.describe(boundVariables)
+			},
+			ifUniversal: {
+				let variables = lazy($0)
+					.map { bound + $0.value.subscriptedDescription }
+					|> sorted
+					|> (join <| ",")
+				return "∀{\(variables)}.\($1.describe(boundVariables.union($0)))"
+			})
```

Comment:
This enables clever printing of types such that bound variables are represented by `α` and free ones by `τ`, and function parameters are parenthesized iff they would be ambiguous.